qcodes/utils/validators.py

```diff
@@ -488,32 +488,72 @@ def __repr__(self) -> str:
 
 class Arrays(Validator):
     """
-    Validator for numerical numpy arrays
+    Validator for numerical numpy arrays of numeric types (int, float, complex).
+    By default it validates int and float arrays.
+
+    Min and max validation is not supported for complex numbers.
 
     Args:
         min_value:  Min value allowed, default inf.
         max_value: Max value allowed, default inf.
         shape: The shape of the array, tuple of either ints or Callables taking
             no arguments that return the size along that dim as an int.
+        valid_types: Sequence of types that the validator should support. Should
+            be a subset of the supported types or None. If None all real
+            datatypes will validate.
     """
 
-    validtypes = (int, float, np.integer, np.floating)
+    real_types = (np.integer, np.floating)
+    supported_types = real_types + (np.complexfloating,)
 
     def __init__(self, min_value: numbertypes = -float("inf"),
                  max_value: numbertypes = float("inf"),
-                 shape: TSequence[shape_type] = None) -> None:
+                 shape: TSequence[shape_type] = None,
+                 valid_types: Optional[TSequence[type]] = None) -> None:
 
-        if isinstance(min_value, self.validtypes):
-            self._min_value = min_value
+        if valid_types is not None:
+            for mytype in valid_types:
+
+                is_supported = any(np.issubsctype(mytype, supported_type) for
+                                   supported_type in self.supported_types)
+                if not is_supported:
+                    raise TypeError(f"Array Validator only supports numeric "
+                                    f"types: {mytype} is not supported.")
+
+            self.valid_types = valid_types
         else:
-            raise TypeError('min_value must be a number')
+            self.valid_types = self.real_types
 
-        valuesok = max_value > min_value
+        supports_complex = any(
+            np.issubsctype(my_type, np.complexfloating) for my_type in
+            self.valid_types)
 
-        if isinstance(max_value, self.validtypes) and valuesok:
+        inf_limits = np.isinf(min_value) and np.isinf(max_value)
+
+        if supports_complex and not inf_limits:
+            raise TypeError("min and max_value is not supported for complex "
+                            "types.")
+
+        if any(np.issubsctype(type(min_value), real_type) for real_type in
+               self.real_types):
+            self._min_value = min_value
+        else:
+            raise TypeError('min_value must be a (real) number and the '
+                            'validator must only allow real (non complex) '
+                            'data types')
+
+        if any(np.issubsctype(type(max_value), real_type) for real_type in
+               self.real_types):
```

Comment:
why is it `real_types` and not `valid_types`? if it's array of integers, should the min/max be allowed to be float (i'd say, no)? if it's array of floats, should min/max be allowed to be integers (i'd say, yes, but that's not very clean)?

Reply:
Because that would be an api change

Reply:
Because min and max cannot be complex. Complex numbers do not have partial order

Reply:
are you saying that previously it was allowed to specify the min/max with value of some type regardless of what `valid_types` are? wow.

Reply:
But sure we can extend it to also check that min and max is of that same type

Reply:
sure. but this (valid_types==complex + min/max are specified) is caught above, right?

My question is why the type of min/max value is being checked against all real_types instead of valid_types of the validator (the `self`)?

Reply:
I don't think so. See comment below

Reply:
> to also check that min and max is of that same type

i think you can catch two hairs at the same time by saying "min and max have to be of that same type as the `valid_types`", would you agree?